### ISSUE TITLE
Pop-up Mapping Stack

### ISSUE THREAD

**Issue by vy-storm** [florisboard/florisboard#2008]
An option to have multiple pop-up mappings stack as seen in gboard.

**Comment by patrickgold** [comment]
Like discussed in the Matrix group chat, I am open to implement this. However we need to find a way so this works relatively smooth and consistent for all languages (at least the ones that use Latin scripts).

The tricky parts already begins with the fact that many popup mappings have duplicate accent letters all sorted in a different order, so here the question already arises, how do I sort the popups once I merged them without having duplicates? There is no factor which allows for this sort of filtering, so the popup logic either has to guess or just leave the order as is.

So my idea is to introduce predefined letter frequencies into FlorisBoard: https://en.wikipedia.org/wiki/Letter_frequency#Relative_frequencies_of_letters_in_other_languages

With a table like this we can weight each accent letter and then sort them spherically around the finger in the extended popup panel. I still need to think more about this and make a more concise draft, but what would you think of this idea?

Also: do we even merge the domains and symbols popup panel (long press letter left of the enter key) and if so how do we do that? For domains the idea is to implement custom domain slots, like known already from the currency set slots, however the symbols still remain.

**Comment by vy-storm** [comment]
> Like discussed in the Matrix group chat, I am open to implement this. However we need to find a way so this works relatively smooth and consistent for all languages (at least the ones that use Latin scripts).
> 
> The tricky parts already begins with the fact that many popup mappings have duplicate accent letters all sorted in a different order, so here the question already arises, how do I sort the popups once I merged them without having duplicates? There is no factor which allows for this sort of filtering, so the popup logic either has to guess or just leave the order as is.
> 
> So my idea is to introduce predefined letter frequencies into FlorisBoard: https://en.wikipedia.org/wiki/Letter_frequency#Relative_frequencies_of_letters_in_other_languages
> 
> With a table like this we can weight each accent letter and then sort them spherically around the finger in the extended popup panel. I still need to think more about this and make a more concise draft, but what would you think of this idea?
> 
> Also: do we even merge the domains and symbols popup panel (long press letter left of the enter key) and if so how do we do that? For domains the idea is to implement custom domain slots, like known already from the currency set slots, however the symbols still remain.

My take for implementation would be simply whichever order the languages are sorted, that's the order the symbols appear. However, all duplicate symbols get merged into one, and those duplicate symbols are given lesser weight with the unique ones per language all kept in the front via greater weight.  My solution is simple but not perfect, however I personally think it would work more than sufficiently.

Having the entire special character set weighted has merit, but this would require setting weights for each new language you wish to add, whereas having them weighted automatically can happen to every latin language at once.  These could of course be used together, manual weighting could be done and automatic weighting be used as a catch-all. What do you mean by spherical around the finger though? The pop-up panel is a plane is it not? Do you mean the first closest then the next three closest then the next six closest ect from the middle?

**Comment by patrickgold** [comment]
> My take for implementation would be simply whichever order the languages are sorted, that's the order the symbols appear. However, all duplicate symbols get merged into one, and those duplicate symbols are given lesser weight with the unique ones per language all kept in the front via greater weight. My solution is simple but not perfect, however I personally think it would work more than sufficiently.

Could work, that's a simple yet effective idea. Still the current popup logic (merging the symbol into the popups, while respecting important accent letters) is already complex, so this is gonna be fun once I tackle this :)

> These could of course be used together, manual weighting could be done and automatic weighting be used as a catch-all.

Yep I think a dual system with the automatic system is the best in the long run, however your idea must be tested beforehand to see if it already suffices.

> What do you mean by spherical around the finger though? The pop-up panel is a plane is it not? Do you mean the first closest then the next three closest then the next six closest ect from the middle?

Exactly. My bad for the lax specification.

**Comment by HiGein** [comment]
I second this.

TL;DR: it would be really great to have sort of "layout merge" feature with keeping common symbols.

There are two main features every layout has.

- Primary shape (example: QWERTZ, AZERTY, DVORAK)
- Language

In fact, there are many examples of people using multiple languages in typing while these languages share the same script. Examples:

- In Poland people might also use Czech or German or Lithuanian. On the other hand, there is Kashubian language in Poland which has some extra diacritics.
- In Andorra people might use Catalan, Spanish and French simultaneously, these languages have different diacritics.
- Eastern slavic languages are often mutually intelligible (partially) and people might want to type both Belarussian and Ukrainian.

There's no much sense having both German and Polish layout simultaneously, since languages share a lot of common letters. On the other hand, people who use several languages mostly have "main" language. What I'm trying to say is, for example, Polish person could have a lot of contacts with Germans and would like to add German letters in addition to usual QWERTY Polish layout or vice versa: German person who would like to add Polish letters in addition to regular German QWERTZ. There might be hardcore cases as well such as adding special diacritics to the COLEMAK.

It's not possible to predict all the cases and redefine all the layouts. On the other hand, while totally custom layouts are cool, they might be an overkill. Also, it's not possible to have selectable all existing diacritics in convenient interface, there's too much of them :)

It would be awesome if it would be possible to sort of "generate" layout based on some features! Let's say I would like QWERTY and Spanish and Czech languages, then the app would allow me to select this and have a usable layout.

Of course there are questions, for example, where to put diacritics for similar letters.

Probably it's better to have kind of "main" layout and possibility to add some symbols there based on the language (obviously, it would not be possible to add cyrillic layour to latin ones).

**Comment by omarcinkonis** [comment]
## Problem

> So my idea is to introduce predefined letter frequencies into FlorisBoard: https://en.wikipedia.org/wiki/Letter_frequency#Relative_frequencies_of_letters_in_other_languages

Ranking diacritics by frequency is not a good idea in my opinion. It will increase mental load, resulting in slower typing.

As an example, consider these Lithuanian diacritics:
- ęė
- ųū

The type of diacritic used in ę and ų is called [ogonek](https://en.wikipedia.org/wiki/Ogonek).

It would be intuitive for ogonek to always be in the same spot:
- If I hold e, I expect to get ę (e ogonek) as the primary option.
- If I hold u, I expect to get ų (u ogonek) as the primary option.

However, ordering by frequency could result in different ogonek positions on different keys:
- Hold e, get ę.
- Hold u, get ū, while ų is on the side.

Speaking from personal experience, the random order of diacritics is the reason why I haven't switched to Florisboard yet. Having diacritics ranked by frequency would still appear "random" (not logical) to me and would not solve the problem.

## Solution

In my opinion, a better solution would be to rank primary, secondary, tertiary <...> diacritics for each language manually, not by frequency.

Example ranks:
- Lithuanian ą, č, ę, ė, į, ų, ū, š, ž
- Polish ą, ć, ę, ł, ń, ó, ś, ź, ż
- Czech á, č, ď, ě, í, ň, ó, ř, ś, ů, ý, ž

If the user has multiple languages with diacritics enabled, list diacritics by primary language rank, then list diacritics by secondary language rank and so on. Do not repeat diacritics that overlap between two languages. Letters e, r, y, u, i, o are in the top row and have numbers. Numbers should have the least priority. 

Example merged Lithuanian primary + Polish secondary + Czech tertiary:
ą á
č ć
ď
ę ė ě 3
ł
ń ň
ó 9
ř 4
į í 8
ý 6
ų ū ů 7
š ś
ž ź ż

Now this would not feel random. If I could have this kind of ordering, I would switch to Florisboard immediately.
